lib/observability/tracing/client.go

```diff
@@ -0,0 +1,214 @@
+	}
+
+	var httpOptions []otlptracehttp.Option
+	grpcOptions := []otlptracegrpc.Option{otlptracegrpc.WithDialOption(grpc.WithBlock())}
+
+	if cfg.TLSConfig != nil {
+		httpOptions = append(httpOptions, otlptracehttp.WithTLSClientConfig(cfg.TLSConfig.Clone()))
+		grpcOptions = append(grpcOptions, otlptracegrpc.WithTLSCredentials(credentials.NewTLS(cfg.TLSConfig.Clone())))
+	} else {
+		httpOptions = append(httpOptions, otlptracehttp.WithInsecure())
+		grpcOptions = append(grpcOptions, otlptracegrpc.WithInsecure())
+	}
+
+	var traceClient otlptrace.Client
+	switch cfg.exporterURL.Scheme {
+	case "http", "https":
+		httpOptions = append(httpOptions, otlptracehttp.WithEndpoint(cfg.Endpoint()))
+		traceClient = otlptracehttp.NewClient(httpOptions...)
+	case "grpc":
+		grpcOptions = append(grpcOptions, otlptracegrpc.WithEndpoint(cfg.Endpoint()))
+		traceClient = otlptracegrpc.NewClient(grpcOptions...)
+	case "file":
+		limit := DefaultFileLimit
+		rawLimit := cfg.exporterURL.Query().Get("limit")
+		if rawLimit != "" {
+			convertedLimit, err := strconv.ParseUint(rawLimit, 10, 0)
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+			limit = convertedLimit
+		}
+
+		client, err := NewRotatingFileClient(cfg.Endpoint(), limit)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		traceClient = client
+	default:
+		return nil, trace.BadParameter("unsupported exporter scheme: %q", cfg.exporterURL.Scheme)
+	}
+
+	return traceClient, nil
+}
+
+var _ otlptrace.Client = (*RotatingFileClient)(nil)
+
+// RotatingFileClient is an otlptrace.Client that writes traces to a file. It
+// will automatically rotate files when they reach the configured limit. Each
+// line in the file is a JSON-encoded otlp.Span.
+type RotatingFileClient struct {
+	dir     string
+	limit   uint64
+	written uint64
+
+	lock   sync.Mutex
+	writer io.WriteCloser
+}
+
+func fileName() string {
+	return fmt.Sprintf("%d-*.trace", time.Now().UTC().UnixNano())
+}
+
+// DefaultFileLimit is the default file size limit used before
+// rotating to a new traces file
+const DefaultFileLimit uint64 = 1048576 * 100 // 100MB
+
+// NewRotatingFileClient returns a new RotatingFileClient that will store files in the
+// provided directory. The files will be rotated when they reach the provided limit.
+func NewRotatingFileClient(dir string, limit uint64) (*RotatingFileClient, error) {
+	if err := os.Mkdir(dir, 0700); err != nil && !errors.Is(err, os.ErrExist) {
+		return nil, trace.ConvertSystemError(err)
+	}
+
+	f, err := os.CreateTemp(dir, fileName())
+	if err != nil {
+		return nil, trace.ConvertSystemError(err)
+	}
+
+	return &RotatingFileClient{
+		dir:    dir,
+		limit:  limit,
+		writer: f,
+	}, nil
+}
+
+// Start is a noop needed to satisfy the otlptrace.Client interface.
+func (f *RotatingFileClient) Start(ctx context.Context) error {
+	return nil
+}
+
+// Stop closes the active file and stops the client.
+func (f *RotatingFileClient) Stop(ctx context.Context) error {
```

Comment:
If `Stop` gets called after an `UploadTraces` that has gotten the previous file near the limit, a subsequent `UploadTraces` might reopen a new file and ignore the `Stop`, and the `otlptrace.Client` contract states that `Stop` can get called (once) concurrently with multiple `UploadTraces` calls.